cli/azd/internal/vsrpc/server.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package vsrpc
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"go.lsp.dev/jsonrpc2"
+)
+
+type Server struct {
+	aspireService      *AspireService
+	environmentService *EnvironmentService
+	serverService      *ServerService
+}
+
+func NewServer(aspireService *AspireService, environmentService *EnvironmentService, serverService *ServerService) *Server {
+	return &Server{
+		aspireService:      aspireService,
+		environmentService: environmentService,
+		serverService:      serverService,
+	}
+}
+
+// upgrader is the websocket.Upgrader used by the server to upgrade each request to a websocket connection.
+var upgrader = websocket.Upgrader{}
+
+// Serve calls http.Serve with the given listener and a handler that serves the VS RPC protocol.
+func (s *Server) Serve(l net.Listener) error {
+	mux := http.NewServeMux()
+
+	mux.Handle("/AspireService/v1.0", s.aspireService)
+	mux.Handle("/ServerService/v1.0", s.serverService)
+	mux.Handle("/EnvironmentService/v1.0", s.environmentService)
+
+	server := http.Server{
+		ReadHeaderTimeout: 1 * time.Second,
+		Handler:           mux,
+	}
+	return server.Serve(l)
+}
+
+// serveRpc upgrades the HTTP connection to a WebSocket connection and then serves a set of named method using JSON-RPC 2.0.
+func serveRpc(w http.ResponseWriter, r *http.Request, handlers map[string]Handler) {
+	c, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Print("upgrade:", err)
+		return
+	}
+	defer c.Close()
+
+	rpcServer := jsonrpc2.NewConn(NewWebSocketStream(c))
+
+	// TODO(ellismg): Need to add support for cancellation here. This may require modifications to the Go method
```

Comment:
We might want to consider using https://pkg.go.dev/golang.org/x/exp/jsonrpc2 here, as that looks better suited for doing the cancellation support via their Prempter interface on the connection.

Reply:
It ended up being easy enough for us to support it with the existing package.